File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.py
import sys
input = sys.stdin.readline

def solve(x1, y1, x2, y2):
    return prefix[x2][y2] - prefix[x1-1][y2] - prefix[x2][y1-1] + prefix[x1-1][y1-1]


N, M = map(int, input().split())
table = [[0 for _ in range(N+1)] for _ in range(N+1)]
prefix = [[0 for _ in range(N+1)] for _ in range(N+1)]

for i in range(1, N+1):
    line = list(map(int, input().split()))
    for j in range(1, N+1):
        table[i][j] = line[j-1]

for i in range(1, N+1):
    for j in range(1, N+1):
        prefix[i][j] = prefix[i-1][j] + prefix[i][j-1] - prefix[i-1][j-1] + table[i][j]

for i in range(M):
    x1, y1, x2, y2 = map(int, input().split())
    print(solve(x1,y1,x2,y2))

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.*;
import java.util.*;

public class Main {

	static int[][] table = new int[1025][1025];
	static int[][] dp = new int[1025][1025];

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;

		st = new StringTokenizer(br.readLine());

		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());

		for (int i = 1; i <= N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 1; j <= N; j++) {
				table[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		initDP(N);

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int x1 = Integer.parseInt(st.nextToken());
			int y1 = Integer.parseInt(st.nextToken());
			int x2 = Integer.parseInt(st.nextToken());
			int y2 = Integer.parseInt(st.nextToken());
			solve(x1, y1, x2, y2);
		}

	}

	private static void solve(int x1, int y1, int x2, int y2) {
		int result = dp[x2][y2] - dp[x1 - 1][y2] - dp[x2][y1 - 1] + dp[x1 - 1][y1 - 1];
		System.out.println(result);
	}

	private static void initDP(int N) {
		for (int i = 1; i <= N; i++) {
			for (int j = 1; j <= N; j++) {
				dp[i][j] = table[i][j] + dp[i - 1][j] + dp[i][j - 1] - dp[i - 1][j - 1];
			}
		}
	}
}
